File: Util/CreadorDocumentos.py
from reportlab.lib import colors
from reportlab.lib.pagesizes import letter, landscape
from reportlab.platypus import SimpleDocTemplate, Table, TableStyle, Paragraph, Spacer
from reportlab.lib.styles import getSampleStyleSheet
import tempfile

class CreadorDocumentos:
    @staticmethod
    def crear_documento_ventas(ventas, configuracion, fecha_inicio, fecha_fin):
        archivo_temporal = tempfile.NamedTemporaryFile(delete=False, suffix=".pdf", prefix="ReporteVentas_")
        ruta_archivo = archivo_temporal.name
        archivo_temporal.close()

        documento = SimpleDocTemplate(ruta_archivo, pagesize=landscape(letter))
        elementos = []
        estilos = getSampleStyleSheet()
        
        ingresos_totales = sum(v.getTotal() for v in ventas)
        numero_ventas = len(ventas)
        promedio_venta = ingresos_totales / numero_ventas if numero_ventas > 0 else 0

        elementos.append(Paragraph(configuracion.getNombre(), estilos['Title']))
        elementos.append(Paragraph(f"Sucursal: {configuracion.getSucursal()}", estilos['Normal']))
        elementos.append(Paragraph(f"RFC: {configuracion.getRfc()}", estilos['Normal']))
        elementos.append(Paragraph(f"Periodo: {fecha_inicio} al {fecha_fin}", estilos['Normal']))
        elementos.append(Spacer(1, 12))
        
        elementos.append(Paragraph(f"Ventas totales: {numero_ventas}", estilos['Normal']))
        elementos.append(Paragraph(f"Ingresos totales: ${ingresos_totales:.2f}", estilos['Normal']))
        elementos.append(Paragraph(f"Promedio por venta: ${promedio_venta:.2f}", estilos['Normal']))
        elementos.append(Spacer(1, 12))

        lista_datos = [["ID Venta", "Fecha", "Hora", "Cajero", "Metodo Pago", "Total", "SKU", "Producto", "Cantidad", "Precio Uni", "Subtotal"]]

        for v in ventas:
            fecha_str = v.getFecha().strftime("%d/%m/%Y") if v.getFecha() else ""
            hora_str = v.getFecha().strftime("%H:%M:%S") if v.getFecha() else ""
            
            if not v.getItems():
                lista_datos.append([
                    str(v.getId()), fecha_str, hora_str, v.getCajero(), v.getMetodoPago(),
                    f"${v.getTotal():.2f}", "", "", "", "", ""
                ])
            else:
                for item in v.getItems():
                    lista_datos.append([
                        str(v.getId()), fecha_str, hora_str, v.getCajero(), v.getMetodoPago(),
                        f"${v.getTotal():.2f}", str(item.getProducto().getId()), 
                        item.getProducto().getNombre(), f"{item.getCantidad():.2f}",
                        f"${item.getProducto().getPrecio():.2f}", f"${item.getSubtotal():.2f}"
                    ])

        if len(lista_datos) == 1:
            lista_datos.append(["", "", "", "", "", "", "", "Sin ventas en el periodo", "", "", ""])

        tabla = Table(lista_datos)
        estilo_tabla = TableStyle([
            ('BACKGROUND', (0, 0), (-1, 0), colors.HexColor("#f4f4f5")),
            ('TEXTCOLOR', (0, 0), (-1, 0), colors.black),
            ('ALIGN', (0, 0), (-1, -1), 'CENTER'),
            ('FONTNAME', (0, 0), (-1, 0), 'Helvetica-Bold'),
            ('BOTTOMPADDING', (0, 0), (-1, 0), 12),
            ('BACKGROUND', (0, 1), (-1, -1), colors.white),
            ('GRID', (0, 0), (-1, -1), 1, colors.black)
        ])
        tabla.setStyle(estilo_tabla)
        elementos.append(tabla)

        documento.build(elementos)

        return ruta_archivo

File: Util/CreadorDocumentos.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using QuestPDF.Fluent;
using QuestPDF.Helpers;
using QuestPDF.Infrastructure;
using punto_de_venta_C_.Modelo;

namespace punto_de_venta_C_.Util
{
    public class CreadorDocumentos
    {
        public static string crearDocumentoVentas(List<Venta> ventas, ConfiguracionTienda configuracion, string fechaInicio, string fechaFin)
        {
            QuestPDF.Settings.License = LicenseType.Community;

            string rutaArchivo = Path.Combine(Path.GetTempPath(), "ReporteVentas_" + Guid.NewGuid().ToString() + ".pdf");

            double ingresosTotales = ventas.Sum(v => v.getTotal());
            int numeroVentas = ventas.Count;
            double promedioVenta = numeroVentas > 0 ? ingresosTotales / numeroVentas : 0;

            Document.Create(documento =>
            {
                documento.Page(pagina =>
                {
                    pagina.Size(PageSizes.Letter.Landscape());
                    pagina.Margin(1, Unit.Centimetre);
                    pagina.PageColor(QuestPDF.Helpers.Colors.White);
                    pagina.DefaultTextStyle(x => x.FontSize(9).FontFamily("Arial"));

                    pagina.Header().Column(col =>
                    {
                        col.Item().Text(configuracion.getNombre()).FontSize(18).SemiBold().FontColor(QuestPDF.Helpers.Colors.Blue.Darken2);
                        col.Item().Text($"Sucursal: {configuracion.getSucursal()}");
                        col.Item().Text($"RFC: {configuracion.getRfc()}");
                        col.Item().Text($"Periodo: {fechaInicio} al {fechaFin}");
                        col.Item().PaddingTop(10).Text($"Ventas totales: {numeroVentas}");
                        col.Item().Text($"Ingresos totales: {ingresosTotales:C}");
                        col.Item().Text($"Promedio por venta: {promedioVenta:C}");
                        col.Item().PaddingBottom(10);
                    });

                    pagina.Content().Table(tabla =>
                    {
                        tabla.ColumnsDefinition(cols =>
                        {
                            cols.ConstantColumn(40);
                            cols.ConstantColumn(50);
                            cols.ConstantColumn(40);
                            cols.ConstantColumn(70);
                            cols.ConstantColumn(60);
                            cols.ConstantColumn(50);
                            cols.ConstantColumn(50);
                            cols.RelativeColumn();
                            cols.ConstantColumn(40);
                            cols.ConstantColumn(50);
                            cols.ConstantColumn(50);
                        });

                        tabla.Header(cabecera =>
                        {
                            cabecera.Cell().Element(EstiloCelda).Text("ID Venta");
                            cabecera.Cell().Element(EstiloCelda).Text("Fecha");
                            cabecera.Cell().Element(EstiloCelda).Text("Hora");
                            cabecera.Cell().Element(EstiloCelda).Text("Cajero");
                            cabecera.Cell().Element(EstiloCelda).Text("Metodo");
                            cabecera.Cell().Element(EstiloCelda).Text("Total");
                            cabecera.Cell().Element(EstiloCelda).Text("SKU");
                            cabecera.Cell().Element(EstiloCelda).Text("Producto");
                            cabecera.Cell().Element(EstiloCelda).Text("Cant");
                            cabecera.Cell().Element(EstiloCelda).Text("Precio");
                            cabecera.Cell().Element(EstiloCelda).Text("Subtotal");

                            static QuestPDF.Infrastructure.IContainer EstiloCelda(QuestPDF.Infrastructure.IContainer contenedor)
                            {
                                return contenedor.DefaultTextStyle(x => x.SemiBold()).PaddingVertical(5).BorderBottom(1).BorderColor(QuestPDF.Helpers.Colors.Black);
                            }
                        });

                        foreach (var v in ventas)
                        {
                            string fecha = v.getFecha().ToString("dd/MM/yyyy");
                            string hora = v.getFecha().ToString("HH:mm:ss");

                            if (v.getItems() == null || v.getItems().Count == 0)
                            {
                                tabla.Cell().Element(EstiloCuerpo).Text(v.getId().ToString());
                                tabla.Cell().Element(EstiloCuerpo).Text(fecha);
                                tabla.Cell().Element(EstiloCuerpo).Text(hora);
                                tabla.Cell().Element(EstiloCuerpo).Text(v.getCajero());
                                tabla.Cell().Element(EstiloCuerpo).Text(v.getMetodoPago());
                                tabla.Cell().Element(EstiloCuerpo).Text(v.getTotal().ToString("C"));
                                tabla.Cell().Element(EstiloCuerpo).Text("");
                                tabla.Cell().Element(EstiloCuerpo).Text("Sin detalles");
                                tabla.Cell().Element(EstiloCuerpo).Text("");
                                tabla.Cell().Element(EstiloCuerpo).Text("");
                                tabla.Cell().Element(EstiloCuerpo).Text("");
                            }
                            else
                            {
                                foreach (var item in v.getItems())
                                {
                                    tabla.Cell().Element(EstiloCuerpo).Text(v.getId().ToString());
                                    tabla.Cell().Element(EstiloCuerpo).Text(fecha);
                                    tabla.Cell().Element(EstiloCuerpo).Text(hora);
                                    tabla.Cell().Element(EstiloCuerpo).Text(v.getCajero());
                                    tabla.Cell().Element(EstiloCuerpo).Text(v.getMetodoPago());
                                    tabla.Cell().Element(EstiloCuerpo).Text(v.getTotal().ToString("C"));
                                    tabla.Cell().Element(EstiloCuerpo).Text(item.getProducto().getId());
                                    tabla.Cell().Element(EstiloCuerpo).Text(item.getProducto().getNombre());
                                    tabla.Cell().Element(EstiloCuerpo).Text(item.getCantidad().ToString("0.00"));
                                    tabla.Cell().Element(EstiloCuerpo).Text(item.getProducto().getPrecio().ToString("C"));
                                    tabla.Cell().Element(EstiloCuerpo).Text(item.getSubtotal().ToString("C"));
                                }
                            }
                        }

                        if (ventas.Count == 0)
                        {
                            tabla.Cell().ColumnSpan(11).PaddingVertical(10).AlignCenter().Text("Sin ventas en el periodo");
                        }

                        static QuestPDF.Infrastructure.IContainer EstiloCuerpo(QuestPDF.Infrastructure.IContainer contenedor)
                        {
                            return contenedor.BorderBottom(1).BorderColor(QuestPDF.Helpers.Colors.Grey.Lighten2).PaddingVertical(3);
                        }
                    });

                    pagina.Footer().AlignCenter().Text(x =>
                    {
                        x.Span("Pagina ");
                        x.CurrentPageNumber();
                        x.Span(" de ");
                        x.TotalPages();
                    });
                });
            })
            .GeneratePdf(rutaArchivo);

            return rutaArchivo;
        }
    }
}
